Building a Reinforcement Learning Environment using OpenAI Gym (#5966)
section-engineering-education/engineering-education: content/articles/building-a-reinforcement-learning-environment-using-openai-gym/index.md

- [Building the custom RL environment with OpenAI Gym](#building-the-custom-rl-environment-with-openai-gym)
- [Building the agent with Keras-RL](#building-the-agent-with-keras-rl)
- [Creating a Deep Learning model using Keras](#creating-a-deep-learning-model-using-keras)
- [Testing our custom RL environment](#testing-our-custom-rl-environment)
- [Wrapping up](#wrapping-up)
- [Further reading](#further-reading)

### Goals
There are a couple of things we need to note before we begin:
- We want our optimal temperature to be between 37 and 39 degrees Celcius. 
- The shower length will be 60 seconds. If you've worked with other standard RL environments, you know that they have an episode length. In this case, our episode length will be 60 seconds. This means that the model will try to get into that optimal temperature range within 60 seconds.
- Our model will be performing three actions; Turn up, leave, and turn down. We can turn our temperature up, turn it down, or leave it as it is. 

Let's begin by installing our dependencies.

### Installing and importing required dependencies
We will be installing four key dependencies:

> - `TensorFlow` allows us to perform training and inferencing of deep learning models.

Reviewer: ```suggestion
`TensorFlow` allows us to perform training and inference on deep learning models.
```

- `OpenAI Gym` to allow us to build our environment.
- `Keras` is a high-level API that allows us to build deep learning models.
- `Keras-rl2` gives us several pre-defined agents to build RL models.

```bash
!pip install tensorflow==2.3.0 gym keras keras-rl2
```
The next step involves importing them into our notebook.

```python
import numpy as np
from gym import Env
from gym.spaces import Box, Discrete
import random
```
We've imported:
- The `Env` class from OpenAI Gym. The placeholder class allows us to build our custom environment on top of it.
- The `Discrete` and `Box` spaces from `gym.spaces`. They allow us to define the actions and the current state we can take on our environment. 
- `numpy` to help us with the math.
- `random` to allow us to test out our random environment.

### Building the custom RL environment with OpenAI Gym
We begin by creating a `CustomEnv` class. When we pass `Env` to the `CustomEnv` class, we inherit the methods and properties from the OpenAI Gym environment class.

```python
class CustomEnv(Env):
```

> We've gone ahead and implemented four different functions within the `CustomEnv` class. We are creating the `__init__` function to initialize the actions, observations, and episode length that we can take when creating your environment. 

Reviewer: ```suggestion
We've gone ahead and implemented four different functions within the `CustomEnv` class. We create the `__init__` function to initialize the actions, observations, and episode length we can take when creating your environment. 
```

`Discrete` spaces take in a fixed range of non-negative values. For our case, it takes three actions; `down (0)`, `stay(1)`, `up (2)`. The `observation_space` will hold an array of our current temperature. Next, we set our start temperature to `38` degrees plus a random integer. Finally, we've set the shower length to `60` seconds.

> Unlike `Discrete` spaces, `Box` spaces are much more flexible and allow you to pass through multiple values between `0` and `100`. In addition, you can use it to hold images, audio, and data frames.

```python
    def __init__(self):
        self.action_space = Discrete(3)
        self.observation_space = Box(low=np.array([0]), high=np.array([100]))
        self.state = 38 + random.randint(-3,3)
        self.shower_length = 60
```
The `step` function defines what we do after we take action. We've set our action value to `-1`. Ideally, this means that:

> - If we apply action (0), and apply `-1`, we get a `-1` value. This action will lower the temperature by `1`.
> - If we apply action (1), and apply `-1`, we get a `0` value. This action will make the temperature stay where it is.
> - If we apply action (2), and apply `-1`, we get a `1` value. This action will increase the temperature by `1`.

Reviewer: ```suggestion
- If we apply action (0) together with `-1`, we get a `-1` value. This action will lower the temperature by `1`.
- If we apply action (1) together with `-1`, we get a `0` value. This action will make the temperature stay where it is.
- If we apply action (2) together with `-1`, we get a `1` value. This action will increase the temperature by `1`.
```

We are also reducing the shower length by `1`.

When calculating the reward, if our temperature is in its optimal range of `37`, and `39`, we give a reward of `1`. If it isn't in this optimal range, give a reward of `-1`. Our model will always try to converge with this function so that the temperature is within the optimal range.

```python
    def step(self, action):
        self.state += action -1 
        self.shower_length -= 1 
        
        # Calculating the reward
        if self.state >=37 and self.state <=39: 
            reward =1 
        else: 
            reward = -1 
        
        # Checking if shower is done
        if self.shower_length <= 0: 
            done = True
        else:
            done = False
        
        # Setting the placeholder for info
        info = {}
        
        # Returning the step information
        return self.state, reward, done, info
```

> The `render` function is used to visualize your results. However, we won't be using it for this tutorial. But, this is where you'd write the visualization code.

Reviewer: Please rewrite this sentence in active voice.
